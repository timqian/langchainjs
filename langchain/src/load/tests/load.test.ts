import { test, expect } from "@jest/globals";
import { stringify } from "yaml";
import { z } from "zod";

import { load } from "../index.js";
import { OpenAI, PromptLayerOpenAI } from "../../llms/openai.js";
import { PromptTemplate } from "../../prompts/prompt.js";
import { LLMChain } from "../../chains/llm_chain.js";
import { Cohere } from "../../llms/cohere.js";
import {
  HumanMessagePromptTemplate,
  SystemMessagePromptTemplate,
  ChatPromptTemplate,
} from "../../prompts/chat.js";
import { ChatOpenAI } from "../../chat_models/openai.js";
import { LangChainTracer } from "../../callbacks/index.js";
import {
  FewShotPromptTemplate,
  LengthBasedExampleSelector,
} from "../../prompts/index.js";
import { initializeAgentExecutorWithOptions } from "../../agents/initialize.js";
import { Calculator } from "../../tools/calculator.js";
import { RequestsGetTool } from "../../tools/requests.js";
import { JsonListKeysTool, JsonSpec } from "../../tools/json.js";
import { AgentExecutor } from "../../agents/executor.js";
import { CommaSeparatedListOutputParser } from "../../output_parsers/list.js";
import { StructuredOutputParser } from "../../output_parsers/structured.js";

test("serialize + deserialize llm", async () => {
  const llm = new PromptLayerOpenAI({
    temperature: 0.5,
    modelName: "davinci",
    openAIApiKey: "openai-key",
    promptLayerApiKey: "promptlayer-key",
  });
  const lc_argumentsBefore = llm.lc_kwargs;
  const str = JSON.stringify(llm, null, 2);
  expect(llm.lc_kwargs).toEqual(lc_argumentsBefore);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  expect(JSON.parse(str).kwargs.temperature).toBe(0.5);
  expect(JSON.parse(str).kwargs.modelName).toBe("davinci");
  expect(JSON.parse(str).kwargs.openAIApiKey.type).toBe("secret");
  expect(JSON.parse(str).kwargs.promptLayerApiKey.type).toBe("secret");
  const llm2 = await load<OpenAI>(str, {
    OPENAI_API_KEY: "openai-key",
    PROMPTLAYER_API_KEY: "promptlayer-key",
  });
  expect(llm2).toBeInstanceOf(OpenAI);
  expect(JSON.stringify(llm2, null, 2)).toBe(str);
});

test("serialize + deserialize llm with optional deps", async () => {
  const llm = new Cohere({ temperature: 0.5, apiKey: "cohere-key" });
  const str = JSON.stringify(llm, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const llm2 = await load<Cohere>(
    str,
    { COHERE_API_KEY: "cohere-key" },
    { "langchain/llms/cohere": { Cohere } }
  );
  expect(llm2).toBeInstanceOf(Cohere);
  expect(JSON.stringify(llm2, null, 2)).toBe(str);
  const llm3 = await load<Cohere>(
    str,
    { COHERE_API_KEY: "cohere-key" },
    { "langchain/llms/cohere": import("../../llms/cohere.js") }
  );
  expect(llm3).toBeInstanceOf(Cohere);
  expect(JSON.stringify(llm3, null, 2)).toBe(str);
});

test("serialize + deserialize llm chain string prompt", async () => {
  const llm = new OpenAI({
    temperature: 0.5,
    modelName: "davinci",
    verbose: true,
    openAIApiKey: "openai-key",
    callbacks: [
      new LangChainTracer(),
      // This custom handler is not serialized
      {
        handleLLMEnd(output) {
          console.log(output);
        },
      },
    ],
  });
  const prompt = PromptTemplate.fromTemplate("Hello, {name}!");
  const chain = new LLMChain({ llm, prompt });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const chain2 = await load<LLMChain>(str, {
    OPENAI_API_KEY: "openai-key",
  });
  expect(chain2).toBeInstanceOf(LLMChain);
  expect(JSON.stringify(chain2, null, 2)).toBe(str);
});

test("serialize + deserialize llm chain chat prompt", async () => {
  const llm = new ChatOpenAI({
    temperature: 0.5,
    modelName: "gpt-4",
    streaming: true,
    azureOpenAIApiKey: "openai-key",
    azureOpenAIApiInstanceName: "openai-instance",
    azureOpenAIApiDeploymentName: "openai-deployment",
    azureOpenAIApiVersion: "openai-version",
    prefixMessages: [
      {
        role: "system",
        content: "You're a nice assistant",
      },
    ],
  });
  const prompt = ChatPromptTemplate.fromPromptMessages([
    SystemMessagePromptTemplate.fromTemplate("You are talking to {name}."),
    HumanMessagePromptTemplate.fromTemplate("Hello, nice model."),
  ]);
  const chain = new LLMChain({ llm, prompt });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const chain2 = await load<LLMChain>(str, {
    AZURE_OPENAI_API_KEY: "openai-key",
  });
  expect(chain2).toBeInstanceOf(LLMChain);
  expect(JSON.stringify(chain2, null, 2)).toBe(str);
});

test("serialize + deserialize llm chain few shot prompt w/ examples", async () => {
  const llm = new OpenAI({
    temperature: 0.5,
    modelName: "davinci",
    openAIApiKey: "openai-key",
    callbacks: [new LangChainTracer()],
  });
  const prompt = new FewShotPromptTemplate({
    examples: [{ yo: "1" }, { yo: "2" }],
    prefix: "You are a nice assistant",
    examplePrompt: PromptTemplate.fromTemplate("An example about {yo}"),
    suffix: "My name is {name}",
    inputVariables: ["yo", "name"],
  });
  const chain = new LLMChain({ llm, prompt });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const chain2 = await load<LLMChain>(str, {
    OPENAI_API_KEY: "openai-key",
  });
  expect(chain2).toBeInstanceOf(LLMChain);
  expect(JSON.stringify(chain2, null, 2)).toBe(str);
});

test("serialize + deserialize llm chain few shot prompt w/ selector", async () => {
  const llm = new OpenAI({
    temperature: 0.5,
    modelName: "davinci",
    openAIApiKey: "openai-key",
    callbacks: [new LangChainTracer()],
  });
  const examplePrompt = PromptTemplate.fromTemplate("An example about {yo}");
  const prompt = new FewShotPromptTemplate({
    exampleSelector: await LengthBasedExampleSelector.fromExamples(
      [{ yo: "1" }, { yo: "2" }],
      { examplePrompt }
    ),
    prefix: "You are a nice assistant",
    examplePrompt,
    suffix: "My name is {name}",
    inputVariables: ["yo", "name"],
  });
  const chain = new LLMChain({ llm, prompt });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  await expect(
    load<LLMChain>(str, {
      OPENAI_API_KEY: "openai-key",
    })
  ).rejects.toThrow(
    'Trying to load an object that doesn\'t implement serialization: {"lc":1,"type":"not_implemented","id":["langchain","prompts","selectors","LengthBasedExampleSelector"]}'
  );
});

test("serialize + deserialize llmchain with output parser", async () => {
  const llm = new OpenAI({
    temperature: 0.5,
    modelName: "davinci",
    openAIApiKey: "openai-key",
    callbacks: [new LangChainTracer()],
  });
  const prompt = PromptTemplate.fromTemplate(
    "An example about {yo} {format_instructions}"
  );
  const outputParser = new CommaSeparatedListOutputParser();
  const chain = new LLMChain({ llm, prompt, outputParser });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const chain2 = await load<LLMChain>(str, {
    OPENAI_API_KEY: "openai-key",
  });
  expect(chain2).toBeInstanceOf(LLMChain);
  expect(JSON.stringify(chain2, null, 2)).toBe(str);
  expect(await chain2.outputParser?.parse("a, b, c")).toEqual(["a", "b", "c"]);
});

test("serialize + deserialize llmchain with struct output parser throws", async () => {
  const llm = new OpenAI({
    temperature: 0.5,
    modelName: "davinci",
    openAIApiKey: "openai-key",
    callbacks: [new LangChainTracer()],
  });
  const prompt = PromptTemplate.fromTemplate(
    "An example about {yo} {format_instructions}"
  );
  const outputParser = new StructuredOutputParser(
    z.object({
      a: z.string(),
    })
  );
  const chain = new LLMChain({ llm, prompt, outputParser });
  const str = JSON.stringify(chain, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  await expect(
    load<LLMChain>(str, {
      OPENAI_API_KEY: "openai-key",
    })
  ).rejects.toThrow(
    'Trying to load an object that doesn\'t implement serialization: {"lc":1,"type":"not_implemented","id":["langchain","output_parsers","structured","StructuredOutputParser"]}'
  );
});

test("serialize + deserialize agent", async () => {
  const llm = new ChatOpenAI({
    temperature: 0,
    modelName: "gpt-4",
    openAIApiKey: "openai-key",
  });
  const executor = await initializeAgentExecutorWithOptions(
    [
      new Calculator(),
      new RequestsGetTool(),
      new JsonListKeysTool(new JsonSpec({ a: "b" })),
    ],
    llm,
    {
      agentType: "chat-conversational-react-description",
    }
  );
  const str = JSON.stringify(executor, null, 2);
  expect(stringify(JSON.parse(str))).toMatchSnapshot();
  const executor2 = await load<AgentExecutor>(
    str,
    { OPENAI_API_KEY: "openai-key" },
    {
      "langchain/tools/calculator": { Calculator },
    }
  );
  expect(executor2).toBeInstanceOf(AgentExecutor);
  expect(JSON.stringify(executor2, null, 2)).toBe(str);
});
