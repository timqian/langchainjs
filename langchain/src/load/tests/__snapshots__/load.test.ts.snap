// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize + deserialize agent 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - agents
  - executor
  - AgentExecutor
kwargs:
  agent:
    lc: 1
    type: constructor
    id:
      - langchain
      - agents
      - chat_convo
      - ChatConversationalAgent
    kwargs:
      llmChain:
        lc: 1
        type: constructor
        id:
          - langchain
          - chains
          - llm_chain
          - LLMChain
        kwargs:
          prompt:
            lc: 1
            type: constructor
            id:
              - langchain
              - prompts
              - chat
              - ChatPromptTemplate
            kwargs:
              inputVariables:
                - chat_history
                - input
                - agent_scratchpad
              promptMessages:
                - lc: 1
                  type: constructor
                  id:
                    - langchain
                    - prompts
                    - chat
                    - SystemMessagePromptTemplate
                  kwargs:
                    lc: 1
                    type: constructor
                    id:
                      - langchain
                      - prompts
                      - prompt
                      - PromptTemplate
                    kwargs:
                      inputVariables: []
                      templateFormat: f-string
                      template: >-
                        Assistant is a large language model trained by OpenAI.


                        Assistant is designed to be able to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics. As a language model, Assistant is able to generate human-like text based on the input it receives, allowing it to engage in natural-sounding conversations and provide responses that are coherent and relevant to the topic at hand.


                        Assistant is constantly learning and improving, and its capabilities are constantly evolving. It is able to process and understand large amounts of text, and can use this knowledge to provide accurate and informative responses to a wide range of questions. Additionally, Assistant is able to generate its own text based on the input it receives, allowing it to engage in discussions and provide explanations and descriptions on a wide range of topics.


                        Overall, Assistant is a powerful system that can help with a wide range of tasks and provide valuable insights and information on a wide range of topics. Whether you need help with a specific question or just want to have a conversation about a particular topic, Assistant is here to assist. However, above all else, all responses must adhere to the format of RESPONSE FORMAT INSTRUCTIONS.
                - lc: 1
                  type: constructor
                  id:
                    - langchain
                    - prompts
                    - chat
                    - MessagesPlaceholder
                  kwargs: chat_history
                - lc: 1
                  type: constructor
                  id:
                    - langchain
                    - prompts
                    - chat
                    - HumanMessagePromptTemplate
                  kwargs:
                    lc: 1
                    type: constructor
                    id:
                      - langchain
                      - prompts
                      - prompt
                      - PromptTemplate
                    kwargs:
                      inputVariables:
                        - input
                      templateFormat: f-string
                      template: >-
                        TOOLS

                        ------

                        Assistant can ask the user to use tools to look up information that may be helpful in answering the users original question. The tools the human can use are:


                        calculator: Useful for getting the result of a math expression. The input to this tool should be a valid mathematical expression that could be executed by a simple calculator.

                        requests_get: A portal to the internet. Use this when you need to get specific content from a website. 
                          Input should be a url string (i.e. "https://www.google.com"). The output will be the text response of the GET request.
                        json_list_keys: Can be used to list all keys at a given path. 
                            Before calling this you should be SURE that the path to this exists.
                            The input is a text representation of the path to the json as json pointer syntax (e.g. /key1/0/key2).

                        RESPONSE FORMAT INSTRUCTIONS

                        ----------------------------


                        Output a JSON markdown code snippet containing a valid JSON object in one of two formats:


                        **Option 1:**

                        Use this if you want the human to use a tool.

                        Markdown code snippet formatted in the following schema:


                        \`\`\`json

                        {{
                            "action": string, // The action to take. Must be one of [calculator, requests_get, json_list_keys]
                            "action_input": string // The input to the action. May be a stringified object.
                        }}

                        \`\`\`


                        **Option #2:**

                        Use this if you want to respond directly and conversationally to the human. Markdown code snippet formatted in the following schema:


                        \`\`\`json

                        {{
                            "action": "Final Answer",
                            "action_input": string // You should put what you want to return to use here and make sure to use valid json newline characters.
                        }}

                        \`\`\`


                        For both options, remember to always include the surrounding markdown code snippet delimiters (begin with "\`\`\`json" and end with "\`\`\`")!



                        USER'S INPUT

                        --------------------

                        Here is the user's input (remember to respond with a markdown code snippet of a json blob with a single action, and NOTHING else):


                        {input}
                - lc: 1
                  type: constructor
                  id:
                    - langchain
                    - prompts
                    - chat
                    - MessagesPlaceholder
                  kwargs: agent_scratchpad
              partialVariables: {}
          llm:
            lc: 1
            type: constructor
            id:
              - langchain
              - chat_models
              - openai
              - ChatOpenAI
            kwargs:
              temperature: 0
              modelName: gpt-4
              openAIApiKey:
                lc: 1
                type: secret
                id:
                  - OPENAI_API_KEY
      outputParser:
        lc: 1
        type: constructor
        id:
          - langchain
          - agents
          - chat_convo
          - ChatConversationalAgentOutputParserWithRetries
        kwargs:
          baseParser:
            lc: 1
            type: constructor
            id:
              - langchain
              - agents
              - chat_convo
              - ChatConversationalAgentOutputParser
            kwargs:
              - calculator
              - requests_get
              - json_list_keys
          outputFixingParser:
            lc: 1
            type: constructor
            id:
              - langchain
              - output_parsers
              - fix
              - OutputFixingParser
            kwargs:
              parser:
                lc: 1
                type: constructor
                id:
                  - langchain
                  - agents
                  - chat_convo
                  - ChatConversationalAgentOutputParser
                kwargs:
                  - calculator
                  - requests_get
                  - json_list_keys
              retryChain:
                lc: 1
                type: constructor
                id:
                  - langchain
                  - chains
                  - llm_chain
                  - LLMChain
                kwargs:
                  llm:
                    lc: 1
                    type: constructor
                    id:
                      - langchain
                      - chat_models
                      - openai
                      - ChatOpenAI
                    kwargs:
                      temperature: 0
                      modelName: gpt-4
                      openAIApiKey:
                        lc: 1
                        type: secret
                        id:
                          - OPENAI_API_KEY
                  prompt:
                    lc: 1
                    type: constructor
                    id:
                      - langchain
                      - prompts
                      - prompt
                      - PromptTemplate
                    kwargs:
                      inputVariables:
                        - instructions
                        - completion
                        - error
                      templateFormat: f-string
                      template: >-
                        Instructions:

                        --------------

                        {instructions}

                        --------------

                        Completion:

                        --------------

                        {completion}

                        --------------


                        Above, the Completion did not satisfy the constraints given in the Instructions.

                        Error:

                        --------------

                        {error}

                        --------------


                        Please try again. Please only respond with an answer that satisfies the constraints laid out in the Instructions:
          toolNames:
            - calculator
            - requests_get
            - json_list_keys
      allowedTools:
        - calculator
        - requests_get
        - json_list_keys
  tools:
    - lc: 1
      type: constructor
      id:
        - langchain
        - tools
        - calculator
        - Calculator
      kwargs: {}
    - lc: 1
      type: constructor
      id:
        - langchain
        - tools
        - RequestsGetTool
      kwargs: {}
    - lc: 1
      type: constructor
      id:
        - langchain
        - tools
        - JsonListKeysTool
      kwargs:
        lc: 1
        type: constructor
        id:
          - langchain
          - tools
          - json
          - JsonSpec
        kwargs:
          a: b
  agentType: chat-conversational-react-description
"
`;

exports[`serialize + deserialize llm 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - llms
  - openai
  - PromptLayerOpenAI
kwargs:
  temperature: 0.5
  modelName: davinci
  openAIApiKey:
    lc: 1
    type: secret
    id:
      - OPENAI_API_KEY
  promptLayerApiKey:
    lc: 1
    type: secret
    id:
      - PROMPTLAYER_API_KEY
"
`;

exports[`serialize + deserialize llm chain chat prompt 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - chat_models
      - openai
      - ChatOpenAI
    kwargs:
      temperature: 0.5
      modelName: gpt-4
      streaming: true
      azureOpenAIApiKey:
        lc: 1
        type: secret
        id:
          - AZURE_OPENAI_API_KEY
      azureOpenAIApiInstanceName: openai-instance
      azureOpenAIApiDeploymentName: openai-deployment
      azureOpenAIApiVersion: openai-version
      prefixMessages:
        - role: system
          content: You're a nice assistant
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - chat
      - ChatPromptTemplate
    kwargs:
      inputVariables:
        - name
      promptMessages:
        - lc: 1
          type: constructor
          id:
            - langchain
            - prompts
            - chat
            - SystemMessagePromptTemplate
          kwargs:
            lc: 1
            type: constructor
            id:
              - langchain
              - prompts
              - prompt
              - PromptTemplate
            kwargs:
              inputVariables:
                - name
              templateFormat: f-string
              template: You are talking to {name}.
        - lc: 1
          type: constructor
          id:
            - langchain
            - prompts
            - chat
            - HumanMessagePromptTemplate
          kwargs:
            lc: 1
            type: constructor
            id:
              - langchain
              - prompts
              - prompt
              - PromptTemplate
            kwargs:
              inputVariables: []
              templateFormat: f-string
              template: Hello, nice model.
      partialVariables: {}
"
`;

exports[`serialize + deserialize llm chain few shot prompt w/ examples 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - llms
      - openai
      - OpenAI
    kwargs:
      callbacks:
        - lc: 1
          type: constructor
          id:
            - langchain
            - callbacks
            - langchain_tracer
            - LangChainTracer
          kwargs: {}
      temperature: 0.5
      modelName: davinci
      openAIApiKey:
        lc: 1
        type: secret
        id:
          - OPENAI_API_KEY
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - few_shot
      - FewShotPromptTemplate
    kwargs:
      examples:
        - yo: "1"
        - yo: "2"
      prefix: You are a nice assistant
      examplePrompt:
        lc: 1
        type: constructor
        id:
          - langchain
          - prompts
          - prompt
          - PromptTemplate
        kwargs:
          inputVariables:
            - yo
          templateFormat: f-string
          template: An example about {yo}
      suffix: My name is {name}
      inputVariables:
        - yo
        - name
"
`;

exports[`serialize + deserialize llm chain few shot prompt w/ selector 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - llms
      - openai
      - OpenAI
    kwargs:
      callbacks:
        - lc: 1
          type: constructor
          id:
            - langchain
            - callbacks
            - langchain_tracer
            - LangChainTracer
          kwargs: {}
      temperature: 0.5
      modelName: davinci
      openAIApiKey:
        lc: 1
        type: secret
        id:
          - OPENAI_API_KEY
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - few_shot
      - FewShotPromptTemplate
    kwargs:
      exampleSelector:
        lc: 1
        type: constructor
        id:
          - langchain
          - prompts
          - selectors
          - LengthBasedExampleSelector
        kwargs:
          examplePrompt:
            lc: 1
            type: constructor
            id:
              - langchain
              - prompts
              - prompt
              - PromptTemplate
            kwargs:
              inputVariables:
                - yo
              templateFormat: f-string
              template: An example about {yo}
        fields:
          examples:
            - yo: "1"
            - yo: "2"
          exampleTextLengths:
            - 4
            - 4
      prefix: You are a nice assistant
      examplePrompt:
        lc: 1
        type: constructor
        id:
          - langchain
          - prompts
          - prompt
          - PromptTemplate
        kwargs:
          inputVariables:
            - yo
          templateFormat: f-string
          template: An example about {yo}
      suffix: My name is {name}
      inputVariables:
        - yo
        - name
"
`;

exports[`serialize + deserialize llm chain string prompt 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - llms
      - openai
      - OpenAI
    kwargs:
      callbacks:
        - lc: 1
          type: constructor
          id:
            - langchain
            - callbacks
            - langchain_tracer
            - LangChainTracer
          kwargs: {}
        - {}
      temperature: 0.5
      modelName: davinci
      verbose: true
      openAIApiKey:
        lc: 1
        type: secret
        id:
          - OPENAI_API_KEY
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - prompt
      - PromptTemplate
    kwargs:
      inputVariables:
        - name
      templateFormat: f-string
      template: Hello, {name}!
"
`;

exports[`serialize + deserialize llm with optional deps 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - llms
  - cohere
  - Cohere
kwargs:
  temperature: 0.5
  apiKey:
    lc: 1
    type: secret
    id:
      - COHERE_API_KEY
"
`;

exports[`serialize + deserialize llmchain with output parser 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - llms
      - openai
      - OpenAI
    kwargs:
      callbacks:
        - lc: 1
          type: constructor
          id:
            - langchain
            - callbacks
            - langchain_tracer
            - LangChainTracer
          kwargs: {}
      temperature: 0.5
      modelName: davinci
      openAIApiKey:
        lc: 1
        type: secret
        id:
          - OPENAI_API_KEY
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - prompt
      - PromptTemplate
    kwargs:
      inputVariables:
        - yo
        - format_instructions
      templateFormat: f-string
      template: An example about {yo} {format_instructions}
  outputParser:
    lc: 1
    type: constructor
    id:
      - langchain
      - output_parsers
      - list
      - CommaSeparatedListOutputParser
    kwargs: {}
"
`;

exports[`serialize + deserialize llmchain with struct output parser throws 1`] = `
"lc: 1
type: constructor
id:
  - langchain
  - chains
  - llm_chain
  - LLMChain
kwargs:
  llm:
    lc: 1
    type: constructor
    id:
      - langchain
      - llms
      - openai
      - OpenAI
    kwargs:
      callbacks:
        - lc: 1
          type: constructor
          id:
            - langchain
            - callbacks
            - langchain_tracer
            - LangChainTracer
          kwargs: {}
      temperature: 0.5
      modelName: davinci
      openAIApiKey:
        lc: 1
        type: secret
        id:
          - OPENAI_API_KEY
  prompt:
    lc: 1
    type: constructor
    id:
      - langchain
      - prompts
      - prompt
      - PromptTemplate
    kwargs:
      inputVariables:
        - yo
        - format_instructions
      templateFormat: f-string
      template: An example about {yo} {format_instructions}
  outputParser:
    lc: 1
    type: not_implemented
    id:
      - langchain
      - output_parsers
      - structured
      - StructuredOutputParser
"
`;
